rpyc/rpyc_client/command_handlers.py

```diff
@@ -0,0 +1,203 @@
+        if len(split_input) < 2:
+            raise TypeError('expecting at least one argument')
+        return {'path_to_executable': split_input[1], 'argv_to_exe': split_input[2:]}
+
+    def _execute(self, path_to_executable, argv_to_exe):
+        remote_subprocess = self.rpyc_conn.modules.subprocess
+        string_args = ' '.join(argv_to_exe)
+        command_line_input = path_to_executable + ' ' + string_args
+        process = remote_subprocess.Popen(command_line_input)
+        return None, process
+
+
+class MonitorHandler(CommandHandler):
+    def _parse_input(self, split_input, additional_input):
+        if len(split_input) != 3:
+            raise TypeError("expecting 2 arguments")
+
+        if split_input[1] == '-r':
+            return {'monitored_path': split_input[2], 'log_path': '', 'monitors': additional_input, 'remove': True}
+
+        return {'monitored_path': split_input[1], 'log_path': split_input[2], 'monitors': additional_input}
+
+    def _execute(self, monitored_path, log_path, monitors, remove=False):
+        if not remove and monitored_path in monitors:
+            return self.empty_output
+
+        if remove and monitored_path in monitors:
+            monitors[monitored_path].stop()
+            del monitors[monitored_path]
+            return self.empty_output
+
+        if remove and monitored_path not in monitors:
+            return self.empty_output
+
+        # not remove and monitored_path not in monitors
+        return '', self.rpyc_conn.root.FileMonitor(monitored_path, self.__get_callback(log_path))
+
+    # Warning: Careful with this closure
+    def __get_callback(self, log_path):
+        def callback(old_stat, new_stat):
+            with open(log_path, 'a+') as log_file:
+                print(datetime.datetime.now(), file=log_file)
+                print('old stat:', file=log_file)
+                print(old_stat, '\n', file=log_file)
+                print('new stat:', file=log_file)
+                print(new_stat, '\n', file=log_file)
+                print('-------------------------\n', file=log_file)
+
+        return callback
+
+
+class RemoveHandler(CommandHandler):
+    def _parse_input(self, split_input, additional_input):
+        if {split_input[1], split_input[2]} != {'-r', '--empty-files'}:
+            raise TypeError("incorrect usage")
+        return {'path': split_input[3]}
+
+    def __get_remove_empty_files_function(self):
+        def remove_empty_files(base_dir_path):
+            import os
+            for path, subdirs, files in os.walk(base_dir_path):
+                for name in files:
+                    full_file_name = os.path.join(path, name)
+                    if os.stat(full_file_name).st_size == 0:
+                        os.remove(full_file_name)
+
+        return remove_empty_files
+
+    def _execute(self, path):
+        teleported_func = self.rpyc_conn.teleport(self.__get_remove_empty_files_function())
+        teleported_func(path)
```

Comment:
this is a note for the exercise:
I think you should make them create a function that deletes a file if it's empty,
and then have another function that uses rpyc_conn.modules.os to walk through the directory, and call this function
I mean, to really use this as a normal function that you can just call by your own with no problems

what do you think?

Reply:
So let's see if I understand correctly.
You mean, like they have this function which they use a lot in their code that happens to delete an empty file. Then, we want them to use this function on the server, in a specific way: go through a directory recursively and delete all the empty files.

Although this is not the case it teaches more about using RPyC right?

Reply:
mm something like that
more like - I want them to really understand that they have a function and every time they call it, something happens in the remote machine, and this is just a regular function for them

it is too different from what you've done tho..